Add basic dashboard views for lesmis

EMcQ-BES opened the pull request:
### Issue #: https://github.com/beyondessential/tupaia-backlog/issues/2501

### Changes:

- Add Dashboard views and entity vitals items

tcaiger commented on packages/lesmis/src/api/queries/useVitalsData.js:
```diff
@@ -0,0 +1,108 @@
+const useDecendantSchoolInformation = (entities, entityCode) => {
+  const decendants = getDecendantsOfType(entities, entityCode, 'school');
+  return useQuery(
+    ['vitals', 'multi_school', entityCode],
```
```suggestion
    ['vitals', 'multiSchool', entityCode],
```

Apart from the rare exception (eg. global constants), it is conventional to use camelCase everywhere in the front end

tcaiger commented on packages/lesmis/src/api/queries/useVitalsData.js:
```diff
@@ -0,0 +1,108 @@
+
+const useDistrictInformation = entityCode => {
+  return useQuery(
+    ['vitals', 'sub_district', entityCode],
```
```suggestion
    ['vitals', 'subDistrict', entityCode],
```

tcaiger commented on packages/lesmis/src/components/EntityVitalsItem.js:
```diff
@@ -0,0 +1,122 @@
+
+const Wrapper = styled.section`
+  padding-top: 0.5rem;
+  padding-left: 0;
+  padding-right: 0;
+  padding-bottom: 0;
```
```suggestion
```

On native html elements padding and margin is 0 by default

tcaiger commented on packages/lesmis/src/components/EntityVitalsItem.js:
```diff
@@ -0,0 +1,122 @@
+  padding-bottom: 0;
+`;
+
+const Container = styled(MuiContainer)`
+  display: flex;
+  flex-direction: row;
+  width: 30%;
+  margin-left: 0;
+  margin-right: 0;
+`;
```
```suggestion

```

The `MuiContainer` is designed as a top level container as it has things like a page gutter and max width 1200px etc. I would also leave the spacing of 30% to the user of this component which will make it more flexible. So then we are just left with the display: flex and flex-direction: row and there is a component for that in the Layout.js file.

tcaiger commented on packages/lesmis/src/components/EntityVitalsItem.js:
```diff
@@ -0,0 +1,122 @@
+
+export const EntityVitalsItem = ({ name, value, icon }) => {
+  return (
+    <Container>
```
```suggestion
    <FlexStart>
```

tcaiger commented on packages/lesmis/src/components/EntityVitalsItem.js:
```diff
@@ -0,0 +1,122 @@
+  }
+};
+
+export const EntityVitalsItem = ({ name, value, icon }) => {
+  return (
```
```suggestion
export const EntityVitalsItem = ({ name, value, icon }) => (
```

tcaiger commented on packages/lesmis/src/components/EntityVitalsItem.js:
```diff
@@ -0,0 +1,122 @@
+  padding-bottom: 0;
+`;
+
+const Container = styled(MuiContainer)`
+  display: flex;
+  flex-direction: row;
+  width: 30%;
+  margin-left: 0;
+  margin-right: 0;
+`;
```
```suggestion
const Container = styled(MuiContainer)`
     .MuiSvgIcon-root {
         padding-top: 0.5rem;
         font-size: 2.5rem;
         color: ${props => props.theme.palette.text.secondary};
       }
`;
```

```suggestion

```

The `MuiContainer` is designed as a top level container as it has things like a page gutter and max width 1200px etc. I would also leave the spacing of 30% to the user of this component which will make it more flexible. So then we are just left with the display: flex and flex-direction: row and there is a component for that in the Layout.js file.

EMcQ-BES replied:
I've swapped these for the actual icons so this formatting is no longer necessary, I've cleared it all out

tcaiger commented on packages/lesmis/src/components/EntityVitalsItem.js:
```diff
@@ -0,0 +1,122 @@
+`;
+
+const StyledPinDrop = styled(PinDrop)`
+  padding-top: 0.5rem;
```
To save repeating these styles on every icon, we can target the icon using its class from the parent component. See the suggestion above.

tcaiger commented on packages/lesmis/src/components/EntityVitalsItem.js:
```diff
@@ -0,0 +1,122 @@
+  color: ${props => props.theme.palette.text.secondary};
+`;
+
+const getIcon = icon => {
```
```suggestion
const VitalsIcon = ({ icon }) => {
```

tcaiger commented on packages/lesmis/src/components/EntityVitalsItem.js:
```diff
@@ -0,0 +1,122 @@
+export const EntityVitalsItem = ({ name, value, icon }) => {
+  return (
+    <Container>
+      {getIcon(icon)}
```
```suggestion
      <VitalsIcon icon={icon} />
```

Best to use a component if possible as its more readable and extensible.

tcaiger commented on packages/lesmis/src/views/DashboardView.js:
```diff
@@ -7,17 +7,149 @@ import React from 'react';
+              <TitleContainer>
+                <RedTitle variant="h4">Province Profile:</RedTitle>
+              </TitleContainer>
+              <EntityVitalsItem name="Province Code" value={vitals.code} icon="PinDrop" />
```
Nice use of entity vitals component. Seems like a good design of component that is flexible and re-usable.

tcaiger commented on packages/lesmis/src/views/DashboardView.js:
```diff
@@ -7,17 +7,149 @@ import React from 'react';
+  height: 100%;
+`;
+
+const WrapContainer = styled(MuiContainer)`
```
```suggestion
const WrapContainer = styled(FlexRow)`
```

Best just to have one MuiContainer at the top level.

EMcQ-BES replied:
From what I can tell `FlexRow` isn't a real type and is just a name used for `styled.div`s being used in a few places, I've swapped `WrapContainer` to be a `styled.div` called `FlexRow`

tcaiger commented on packages/lesmis/src/views/DashboardView.js:
```diff
@@ -7,17 +7,149 @@ import React from 'react';
-);
+const Container = styled(MuiContainer)`
+  margin-left: 0;
+  padding: 0, 0, 0, 0;
```
```suggestion
  padding: 0;
```
padding shortcut. This will set padding to zero on all sides. You may not need to reset the padding though once you re-organise the MuiContainers

tcaiger commented on packages/lesmis/src/views/DashboardView.js:
```diff
@@ -7,17 +7,149 @@ import React from 'react';
+  height: 100%;
+`;
+
+const TitleContainer = styled(MuiContainer)`
+  margin-left: 0;
+  padding-left: 0;
```
```suggestion
const TitleContainer = styled.div)`
```

tcaiger commented on packages/lesmis/src/views/DashboardView.js:
```diff
@@ -7,17 +7,149 @@ import React from 'react';
+export const DashboardView = () => {
+  const { entityCode } = useUrlParams();
+  const vitals = useVitalsData(entityCode);
+  switch (vitals.type) {
```
```suggestion
return (
        <Wrapper>
          <Container>
            <WrapContainer>
               {switch (vitals.type) {
```
I'd move the wrapper components outside of the switch so that you don't have to repeat them.

EMcQ-BES replied:
We attempted this together but had some struggles, if you don't mind I'm going to leave it as is for now and will try clean this up again after I've consolidated my branches a bit

tcaiger commented on packages/lesmis/src/views/DashboardView.js:
```diff
@@ -7,17 +7,149 @@ import React from 'react';
+  }
+};
+
+// <Typography>${JSON.stringify(vitals)}</Typography>
```
```suggestion

```

tcaiger commented on packages/lesmis/src/api/queries/useVitalsData.js:
```diff
@@ -0,0 +1,108 @@
+  return entity.childCodes.map(c => getDecendantsOfType(entities, c, type)).flat();
+};
+
+const useSchoolInformation = entityCode => {
+  return useQuery(
```
```suggestion
const useSchoolInformation = entityCode => useQuery(
```

tcaiger commented on packages/lesmis/src/api/queries/useVitalsData.js:
```diff
@@ -0,0 +1,108 @@
+import { useEntityData } from './useEntityData';
+import { useEntitiesData } from './useEntitiesData';
+
+const getDecendantsOfType = (entities, rootEntityCode, type) => {
```
I would have expected that this kind of aggregation would happen in the report building? Do you know if that's possible?

EMcQ-BES replied:
Not currently, gonna be talking to Rohan about that later today.

EMcQ-BES replied:
Entity aggregation in the report server is coming down the line, but for MVP we're stuck doing it on the front end

tcaiger commented on packages/lesmis/src/api/queries/useVitalsData.js:
```diff
@@ -0,0 +1,108 @@
+};
+
+export const useVitalsData = entityCode => {
+  const { data: entityData } = useEntityData({ entityCode });
+  const { data: entities = [] } = useEntitiesData();
```
```suggestion
  const { data: entities = [], ...entityQuery } = useEntitiesData();
  const entityData = entities.find(e => e.code === entityCode);
```
I would get the current entity from the entities data to potentially save a query and pull out the rest of the react query object to be returned later in `entityQuery`

tcaiger commented on packages/lesmis/src/api/queries/useVitalsData.js:
```diff
@@ -0,0 +1,108 @@
+  const { data: districtData } = useDistrictInformation(entityCode);
+  const { data: subSchoolsData } = useDecendantSchoolInformation(entities, entityCode);
+  const { data: provinceData } = useDecendantDistrictInformation(entities, entityCode);
+  return {
```
```suggestion
  return {
  ...entitiesQuery
```

tcaiger reviewed (changes requested):
Sorry to be a pain with so many suggestions. It's mostly just conventions based stuff which is to be expected with your first big front end task.

EMcQ-BES commented:
> Sorry to be a pain with so many suggestions. It's mostly just conventions based stuff which is to be expected with your first big front end task.

No pain at all, I was wanting to do a pretty big overhaul of this anyway

EMcQ-BES commented:
I've incorporated most of your recommended changes and responded to your comments in places where I have gone a different direction.

I've also cleaned out a lot of unnecessary formatting in my various styled types

tcaiger reviewed (approved):
Hey, Nice work - looking much better. It would be good to clean up that switch at some point and to optimise the queries using the enabled option. I am happy for them to be out of scope of this PR though and maybe we can pair programme them after you've had a chance to consolidate your branches.
